### DISCUSSION TITLE
ACM算法汇总

### DISCUSSION THREAD

**Discussion by 1205240810** [CSGrandeur/s-1problem1day1ac#102] [Algorithm]
# ACM算法汇总

关于所有算法的汇总，大家可以在网上看到很多种不同的分法，他们也都有着自己的作用和意义，但是其中大部分都讲算法按类型划分，没有对难度进行过多的划分。所以本篇文章讲尝试用一种新的方式对这些算法进行分类。实际上。我把整个算法分成了3个部分。入门篇，基础篇和进阶篇。对于算法的学习路径，一直有很多争议。在这里我也只是给出我自己对与算法学习路径的一个总结和归纳。
我认为算法学习的过程应该从广到深。即先把每种算法的基础部分全部学会，然后选择某个算法专心钻研，达到很深的程度。木桶效应告诉我们要补足短处，但是也要有自己的长处。废话不多说，下面就开始对所有算法进行划分和介绍。限于本文篇幅有限，大多算法只会涉及名字。不会深入介绍。
## 入门篇
在基础篇我们会学习很多有用的知识，下面给出一份比较推荐的学习的路径，大家可以根据自己的情况去查缺补漏。
1.枚举
2.模拟
3.递归
4.前缀和和差分
5.尺取法
6.排序算法原理的理解
7.二分，三分
8.分治
9.贪心
10.简单搜索
上面十个算法是比较广泛的算法，很多思想在后续算法的学习中还会起到重要的作用，希望大家都能熟练掌握上面的算法。
## 基础篇
学会了上面的算法后。我们就有了很好的基础帮我们去理解学习下一步的各种算法。在此我一共把算法大概划分成了4个大类，
1.数据结构
2.动态规划
3.图论
4.数学
其中计算几何和博弈论归为数学部分，字符串相关算法归为数据结构。接下来我们对每个部分的基础算法进行了解。
### 1.基础数据结构
对与基础数据结构，很多都是借助STL实现的，所以希望大家都能熟练掌握STL的用法。
1.栈
2.队列
3.链表
4.堆或优先队列
5.HASH表
6.二叉树
7.Trie树
8.set和map的用法
9.并查集
10.KMP算法
11.单调栈
12.树状数组
13.线段树
上述十二种数据结构都能帮助大家更清晰的学习算法并且锻炼代码能力。并且在后续学习中也会继续用到
### 2.基础动态规划
对于动态规划。很多人都感到头疼。所以在基础动态规划部分我们只给出几个基础的DP让大家学习
1.路径搜索dp
2.背包dp
3.区间dp
4.树形dp
5.状态压缩dp
6.概率dp
### 3.基础图论
关于图论的学习，也不用操之过急，从基础入手，锻炼对图论的理解就好
1.建图方法
2.用DFS和BFS遍历整个图
2.欧拉回路
3.拓扑排序
4.三种最短路算法（Fload，Dijkstra，SPFA）
5.两种最小生成树（Prim，Kruskal）
### 4.基础数学
在本次集训中，我们其实没有涉及任何一个数学问题，因为数学的学习不是能直接一步到位的，它更多的要依靠大家的积累。
1.位运算
2.快速幂
3.线性素数筛（埃式筛，欧拉筛）
4.高精度加减乘除
5.组合数学（鸽巢原理，容斥原理等）
6.矩阵快速幂
7.gcd和lcm和exgcd（扩展欧几里得算法）
8.费马小定理
9.乘法逆元

如果上面的算法你都掌握，理解并且能够熟练运用，那么恭喜你，你已经完成了基础组部分算法的学习。之后你就可以愉快的进入进阶组学习更有趣的算法了。
## 进阶篇
进阶篇的算法没有设置任何限制。所以可能涉及算法都会尽量提及。大家也可以按自己的情况来进行学习和练习
### 1.进阶数据结构
1.点分治
2.二叉查找树（Treap，spaly）
3.AC自动机
4.后缀数组
5.后缀树和后缀自动机
6.扩展KMP
7.回文树
8.A*算法和IDA*算法
9.双向BFS
10.AVL树和替罪羊树
11.红黑树
12.树链剖分
13.最近公共祖先（LCA）
14.树上启发式合并（dsu on tree）
15.哈夫曼树
16.ST表
17.主席树
18.划分树
19.莫队算法
20.树套树（线段树套线段树，平衡树套线段树，线段树套平衡树，树状数组套主席树，分块套树状数组）
上面的数据结构几乎囊括了所有竞赛会涉及的数据结构，不管是难度还是数量，都没非常多。大家可以一步步慢慢学习这些数据结构的使用方式
### 2.进阶动态规划
1.DAG上dp
2.数位dp
3.计数dp
4.动态dp
5.插头dp
6.用单调队列或单调栈优化的dp、
7.斜率优化dp
8.四边形不等式优化dp
9.树上dp
### 3.进阶图论
1.k短路
2.强连通分量
3.割点和桥
4.二分图匹配
5.2-SAT
6.基环树
7.图的着色
8.最大流
9.最小割
10.费用流
11.哈密顿图和哈密顿回路
12.弦图
### 4.进阶数学
对于进阶数学我们把他分成数学，计算几何，和博弈论
#### 1.数学部分
1.线性同余方程
2.中国剩余定理
3.二次剩余
4.卢卡斯定理
5.杜教筛
6.莫比乌斯反演
7.Min——25筛
8.原根
9.拉格朗日插值
10.快速傅里叶变换
11.快速数论变换
12.多项式变化
13.牛顿迭代
14.高斯消元
15.调和
16.构造
17.线性规划
18.多项式求根
19.母函数
20.群论
21.定积分的求解
#### 2.计算几何部分
1.叉乘点乘
2.坐标离散化
3.求线段交点
4.欧式距离，曼哈顿距离，和切比雪夫距离
5.皮克定理
6.凸包
7.扫描线算法
8.旋转卡壳
9.平面最近点对
10.平面几何知识
#### 3.博弈论部分
1.巴什博弈
2.威佐夫博弈论
3.Nim游戏
4.斐波那契博弈
5.公平组合博弈
6.SG函数
上面提到了很多算法，他们基本包括个各种各样的算法，但是在算法竞赛中，不只有这些算法题目要掌握，我们还要学习做一些思维题和构造题。如果大家能全部掌握上面所有算法，相信大家也能取得自己想要的成绩了。希望大家不要停下脚步。加油努力的继续在算法学习的道路上越走越远吧！

**Comment by MELEEYIN** [comment]
！？强强？！
